### PR TITLE
Stress test

### PR DESCRIPTION
**Pull request by konradkonrad**
This adds `go test`-runnable testcases to stress- and edge case-test a shutter on gnosis chain live system, in an end-to-end fashion. It does however not use the `encrypting-rpc-server`, but instead submits transactions directly to the sequencer contract.

As a pre-requisite, an account with sufficient `"ETH"` (or other gas tokens corresponding to the chain the system is deployed on) is required. Its private key needs to be available hex encoded in an environment variable `STRESS_TEST_PK=0xcaffee…`.

The most notable test cases are
- `TestStressManyNoWait`, which sends a large number of shutterized transfers at once, in order to test the limits of decryption key generation and propagation on the `keyper` side, as well as decryption performance on the validator side.
- `TestInception`, which sends shutterized transactions inside shutterized transactions.
- `TestStressExceedEncryptedGasLimit`, which tests that there is a limit to the gas that can be used in encrypted transactions per block.

Most other test cases are very simple, and could be considered basic functionality "smoke tests".

**Note**: Due to the live nature of the test environment, not all `PASS`ing tests can show the absence of errors. For example:

`TestStressExceedEncryptedGasLimit`
1) The test relies on a hardcoded value for the limit. If the parameters of the live system change, that value may be outdated.
2) The test shall `PASS` if not all transactions end in the same shutterized block. It can however happen, that an external entity also submitted transactions and therefore the assumed correct behavior happened only by chance.

Also: Since we have no insight to the validator, we rely on timeouts for waiting for transactions to be included on-chain. Additionally, we can not reliably check for failing transactions.

In short: this PR is to some extent ab-using the `go test` environment to allow us excecuting these tests.

There are also two special "Test…" functions, that do not test anything, but can be used for house keeping:
- `TestEmptyAccounts`, which allows to drain all funds from extra accounts that got generated during previous test runs.
- `TestFixNonce`, which can help to clear out nonce-gaps from the public mem-pool for the main test account.

### CODE REVIEW

**Comment by fredo** [comment]
> Haven't reviewed yet, but I think this should be in a separate repository. Are you using a lot of code from the encrypting rpc server that would make this difficult?

Was going to ask the same. Does it make sense to put it into rolling shutter?

File: src/test/stress/stress_test.go [hunk 1]
```diff
@@ -0,0 +1,792 @@
...
+	if err != nil {
+		return *setup, fmt.Errorf("could not query chainId %v", err)
+	}
+	setup.ChainID = chainID
+
+	signerForChain := types.LatestSignerForChainID(chainID)
+	setup.SignerForChain = signerForChain
+
+	submitKeyHex := os.Getenv("STRESS_TEST_PK")
+	if len(submitKeyHex) < 64 {
```
Comment (fredo) [inline src/test/stress/stress_test.go:69]: ooc, why is this condition checking if the ENV variable exists?

File: src/test/stress/stress_test.go [hunk 2]
```diff
@@ -0,0 +1,792 @@
...
+
+	eonKey := &shcrypto.EonPublicKey{}
+	if err := eonKey.Unmarshal(eonKeyBytes); err != nil {
+		return 0, nil, fmt.Errorf("could not unmarshal eonKeyBytes %v", err)
+	}
+	return eon, eonKey, nil
+
+}
+
+func createIdentity() (shcrypto.Block, error) {
```
Comment (fredo) [inline src/test/stress/stress_test.go:315]: nitpick: could you name it `createIdentityPrefix`?

File: src/test/stress/stress_test.go [hunk 3]
```diff
@@ -0,0 +1,809 @@
...
+	setup, err := createSetup(true)
+	if err != nil {
+		log.Fatal("could not create setup", err)
+	}
+	env, err := createStressEnvironment(context.Background(), setup)
+	if err != nil {
+		log.Fatal("could not set up environment", err)
+	}
+
+	env.TransactGasLimitFn = func(value *big.Int, data []byte, toAddress *common.Address, i, count int) uint64 {
```
Comment (fredo) [inline src/test/stress/stress_test.go:577]: Is it intended that `value` iosnt used and the limit is hardcoded?

**Review by fredo — APPROVED** [review]
Looks good in general. Let's discuss where we can put this

**Comment by konradkonrad** [comment]
> Haven't reviewed yet, but I think this should be in a separate repository. Are you using a lot of code from the encrypting rpc server that would make this difficult?

I agree, and it should not be hard, I think the only bit of re-use that's left in there is `rpc.ComputeIdentity`.

File: src/test/stress/stress_test.go (hunk 3 above)
Comment (konradkonrad) [reply src/test/stress/stress_test.go:577]: When I designed the interface
https://github.com/shutter-network/encrypting-rpc-server/pull/13/files#diff-2fde392c0c13754486fec812f19b966baf42611de448feb45f4d1348c1d8b03fR62
I imagined test cases, where we want to variate gas limit depending on the `value`. But it is probably YAGNI. I will remove the argument.

File: src/test/stress/stress_test.go (hunk 1 above)
Comment (konradkonrad) [reply src/test/stress/stress_test.go:69]: go docs
```go
func os.Getenv(key string) string
/*
Getenv retrieves the value of the environment variable named by the key. 
It returns the value, which will be empty if the variable is not present. 
To distinguish between an empty value and an unset value, use LookupEnv.
*/
```
seemed to be sufficient for this use case.

**Comment by konradkonrad** [comment]
This is created elsewhere as requested in https://github.com/shutter-network/encrypting-rpc-server/pull/13#issuecomment-2173968148 and https://github.com/shutter-network/encrypting-rpc-server/pull/13#issuecomment-2175819737
